lua/spectre_common.lua

```diff
@@ -105,25 +129,40 @@ local function determine_if_cacheable(url, namespace, request_headers)
                 refresh_cache = false,
             }
 
-            -- Only cache GET and HEAD requests
-            local http_method = ngx.req.get_method()
-            if http_method ~= 'GET' and http_method ~= 'HEAD' then
-                cacheability_info.is_cacheable = false
-                cacheability_info.reason = 'non-cacheable-method'
-            end
-
             -- Check if client sent no-cache header
             if has_marker_headers(request_headers, PLEASE_DO_NOT_CACHE_HEADERS) then
                 cacheability_info.is_cacheable = false
                 cacheability_info.reason = 'no-cache-header'
                 cacheability_info.refresh_cache = true
             end
+
+            if http_method == 'POST' then
+                if not has_marker_headers(request_headers, SUPPORTED_POST_ENCODING) then
+                    -- For Post requests check the content type is application/json
+                    cacheability_info.is_cacheable = false
+                    cacheability_info.reason = 'non-cacheable-content-type'
+                    cacheability_info.refresh_cache = false
+                elseif cacheability_info.cache_entry.bulk_support then
+                    -- Stop caching if bulk support is added for a POST endpoint.
+                    cacheability_info.is_cacheable = false
+                    cacheability_info.reason = 'no-bulk-support-for-post'
+                    cacheability_info.refresh_cache = false
+                else
+                    -- Start reading the request body into ngx cache.
+                    ngx.req.read_body()
+                    -- For a POST method id extraction and vary fields are obtained from body.
+                    if ngx.var.request_body == nil and (
+                        cache_entry.enable_id_extraction or cache_entry.vary_body_field_list ~= nil
+                    ) then
+                        cacheability_info.is_cacheable = false
+                        cacheability_info.reason = 'non-cacheable-missing-body'
```

Comment:
Could be:
* a missing body (`ngx.var.request_body == nil`)
* `cache_entry.enable_id_extraction` set to `true`
* `cache_entry.vary_body_field_list` set to something else than `nil`

Worth distinguishing between these 3 cases! (`non-cacheable-missing-body`, `non-cacheable-cannot-extract-id-on-post`, `non-cacheable-vary-headers-specified-for-post`)